File: queues/redis/Scala/src/main/scala/com/coredf/redis/RedisClient.scala
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.coredf.redis

import redis.clients.jedis.Jedis

object RedisClient:

  private def env(k: String): String = Option(System.getenv(k)).getOrElse("")
  private def env(k: String, d: String): String = val v = env(k); if v.isEmpty then d else v
  private def envOr(k: String, d: String): String = env(k, d)

  private def encode(value: Any): Array[Byte] = value match
    case null => Array.emptyByteArray
    case b: Array[Byte] => b
    case s: String => s.getBytes(java.nio.charset.StandardCharsets.UTF_8)
    case other => JsonUtil.stringify(other).getBytes(java.nio.charset.StandardCharsets.UTF_8)

  private def decode(raw: Array[Byte]): Any =
    if raw == null then null
    else
      try JsonUtil.parse(new String(raw, java.nio.charset.StandardCharsets.UTF_8))
      catch case _: Exception => new String(raw, java.nio.charset.StandardCharsets.UTF_8)

  private def url(): String =
    val u = env("REDIS_URL")
    if u.nonEmpty then u
    else
      val host = env("REDIS_HOST")
      if host.isEmpty then ""
      else
        val pw = env("REDIS_PASSWORD"); val db = envOr("REDIS_DB", "0")
        if pw.nonEmpty then s"redis://:$pw@$host:${envOr("REDIS_PORT", "6379")}/$db"
        else s"redis://$host:${envOr("REDIS_PORT", "6379")}/$db"

  def Init(): Result = if url().isEmpty then Result.missingEnv("REDIS_URL or REDIS_HOST") else Result.ok()

  def Push(queue: String, value: Any): Result =
    if url().isEmpty then Result.missingEnv("REDIS_URL or REDIS_HOST")
    else
      try val j = new Jedis(url()); try j.lpush(queue.getBytes, encode(value)) finally j.close(); Result.ok()
      catch case e: Exception => Result.transportError(e.getMessage)

  def Pop(queue: String, timeoutSec: Double = 30.0, maxMessages: Int = 1): Result =
    if url().isEmpty then Result.missingEnv("REDIS_URL or REDIS_HOST")
    else
      try
        val messages = scala.collection.mutable.ListBuffer[Map[String, Any]]()
        val j = new Jedis(url())
        try
          var remaining = math.max(1, maxMessages); var deadline = timeoutSec
          while remaining > 0 && deadline > 0 do
            val wait = if remaining == maxMessages then math.max(1, timeoutSec.toInt) else 1
            val item = j.brpop(wait, queue)
            if item == null || item.size() < 2 then ()
            else
              messages += Map("queue" -> queue, "value" -> decode(item.get(1)))
              remaining -= 1; deadline -= wait
        finally j.close()
        Result.ok(Map("messages" -> messages.toList))
      catch case e: Exception => Result.transportError(e.getMessage)

File: queues/redis/Java/src/main/java/com/coredf/redis/RedisClient.java
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.coredf.redis;
import redis.clients.jedis.Jedis; import java.util.*;

public final class RedisClient {
    private RedisClient() {}
    private static String url() {
        String u = MsgUtil.env("REDIS_URL"); if (!u.isEmpty()) return u;
        String host = MsgUtil.env("REDIS_HOST"); if (host.isEmpty()) return "";
        String pw = MsgUtil.env("REDIS_PASSWORD"); String db = MsgUtil.envOr("REDIS_DB", "0");
        if (!pw.isEmpty()) return "redis://:" + pw + "@" + host + ":" + MsgUtil.envOr("REDIS_PORT", "6379") + "/" + db;
        return "redis://"+host+":"+MsgUtil.envOr("REDIS_PORT","6379")+"/"+db;
    }
    public static Result Init() { if (url().isEmpty()) return Result.missingEnv("REDIS_URL or REDIS_HOST"); return Result.ok(); }
    public static Result Push(String queue, Object value) {
        if (url().isEmpty()) return Result.missingEnv("REDIS_URL or REDIS_HOST");
        try (Jedis j = new Jedis(url())) { j.lpush(queue.getBytes(), MsgUtil.encode(value)); return Result.ok(); }
        catch (Exception e) { return Result.transportError(e.getMessage()); }
    }
    public static Result Pop(String queue, double timeoutSec, int maxMessages) {
        if (url().isEmpty()) return Result.missingEnv("REDIS_URL or REDIS_HOST");
        List<Map<String, Object>> messages = new ArrayList<>();
        try (Jedis j = new Jedis(url())) {
            int remaining = Math.max(1, maxMessages); double deadline = timeoutSec;
            while (remaining > 0 && deadline > 0) {
                int wait = remaining == maxMessages ? Math.max(1, (int)timeoutSec) : 1;
                List<byte[]> item = j.brpop((int)wait, queue); if (item == null || item.size() < 2) break;
                messages.add(Map.of("queue", queue, "value", MsgUtil.decode(item.get(1)))); remaining--; deadline -= wait;
            }
        } catch (Exception e) { return Result.transportError(e.getMessage()); }
        return Result.ok(Map.of("messages", messages));
    }
    public static Result Pop(String queue) { return Pop(queue, 30, 1); }
}
